### PR TITLE
feat(mcp): agent-native error envelope, confirm guardrail, and human-approved feedback channel

### PR DESCRIPTION
**Pull request by jan-kubica**
## Why

The MCP server and CLI are consumed almost exclusively by AI agents. This PR makes that first-class in two ways: machine-actionable guardrails on the existing surface, and a feedback path that lets an agent file an issue with the maintainers — gated on explicit human approval and hard sanitization, never on its own.

## Agent guardrails and ergonomics

- **Structured tool-error envelope.** Every MCP tool error now carries `{ error: { code, message, hint?, retryable? } }` with a closed code set (`apps/api/src/mcp/error-codes.ts`): `validation_error`, `missing_scope`, `feature_disabled`, `not_found`, `confirmation_required`, `rate_limited`, `unknown_tool`, `internal_error`. Hints state the next action (which scope to re-consent, closest tool names on typos, retry-after on rate limits).
- **CLI exit codes go live.** The CLI maps envelope codes to its documented exit codes, activating the previously dormant `feature_disabled` (5) and `not_found` (6) paths, and prints `error:` / `hint:` lines on stderr (stdout stays clean for piped JSON).
- **Destructive confirm gate.** Tools with `destructiveHint` refuse without `confirm: true` (`confirmation_required`); `manage_organization` gets an action-level gate for `remove_member`. The CLI injects the confirm only after its own `--yes`/TTY confirmation, so agent-driven deletes always pass through an explicit approval step.
- **Server `instructions`.** Per-mode agent guidance (pagination, text windowing, error envelope, confirm semantics) served at MCP initialize, with char-budget tests.

## Feedback channel (`send_feedback`, `stella feedback send`)

- All content is sanitized server-side before anything is shown or sent: emails, UUIDs, JWT/secret-shaped blobs, IPs, and non-allowlisted URLs are redacted.
- `channel: github` (default, preferred): returns the sanitized preview plus a prefilled new-issue URL and a `gh issue create` one-liner. Nothing is published until the human submits it under their own GitHub account; approval is intrinsic to the flow.
- `channel: email` / `channel: stella`: for humans without a GitHub account. Both require a two-phase handshake — the first call returns the sanitized preview and an HMAC confirmation token bound to that exact content (15-minute TTL); only a second call with the token delivers. `email` goes to the server's configured `FEEDBACK_EMAIL_TO`; `stella` forwards to the hosted intake for instances with no email configured.
- Deliveries are rate-limited to 5/hour per organization; previews and the github channel are uncounted.

## Public intake (`POST /public/feedback`)

Unauthenticated by design (the sender may have no account); protection is abuse bounding, not identity: strict raw-body validation with unknown-key rejection, server-side re-sanitization, 5/hour per-IP rate limit, and 24-hour content dedup. Delivery is email-only (`FEEDBACK_EMAIL_TO`), 503 `feature_disabled` when unset.

## Wiring

- Registry grows to 41 tools (`send_feedback`, scope `stella:feedback` with consent label in all 14 languages, excluded from the anonymized surface); coverage guard and registry-quality ceilings updated deliberately.
- Chat code-mode: `send_feedback` is registered in the write-tool maps but `chatProjectable: false` (same pattern as `fill_template`) — it carries its own approval handshake.
- CLI codegen refreshed (route map, registry snapshot, SKILL.md conventions for errors/confirm/feedback). CLI version unchanged; no publish is triggered.

## Testing

319 apps/api MCP + feedback tests, 148 chat drift-guard tests, 212 CLI tests, MCP coverage guard, i18n checks, apps/api typecheck clean.

## Summary by CodeRabbit

* **New Features**
  * Added an unauthenticated product feedback intake (`POST /public/feedback`) with abuse rate limiting, deduplication, and sanitised email delivery.
  * Introduced the MCP `send_feedback` tool, including a token-based approval flow for email/hosted intake.
  * Added `stella:feedback` consent scope support across the web app and CLI, plus matching translations.
* **Bug Fixes**
  * Improved MCP error handling with stable structured `error.code` envelopes and consistent `not_found`/`internal_error` mapping.
* **Documentation / Tests**
  * Updated MCP/CLI documentation and added extensive automated tests for feedback, confirmations, sanitisation, and structured error envelopes.

### CODE REVIEW

File: apps/api/src/handlers/feedback/intake-guards.ts [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+const cleanupDedupFallback = (fallback: Map<string, number>, now: number) => {
+  if (fallback.size < FALLBACK_CLEANUP_THRESHOLD) {
+    return;
+  }
+  for (const [key, expiresAt] of fallback) {
+    if (expiresAt <= now) {
+      fallback.delete(key);
+    }
+  }
+};
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/feedback/intake-guards.ts:328]: Accepted. The in-memory fallback cleanup now uses per-guard cleanup state and throttles threshold scans to once per minute, so a full but mostly-live fallback map cannot force an O(N) scan on every new claim.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.ts [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+  }
+  for (let keep = composedBody.length; keep > 0; keep -= 128) {
+    const candidate = buildGithubIssueUrl({
+      body: composedBody.slice(0, keep) + GITHUB_BODY_TRUNCATION_MARKER,
+      title,
+    });
+    if (candidate.length <= MAX_GITHUB_ISSUE_URL_CHARS) {
+      return candidate;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.ts:232]: Accepted. The bounded GitHub issue URL path now slices through a helper that drops a dangling high surrogate at the truncation edge, with a regression test for that invariant.

CC on behalf of @jan-kubica

File: packages/cli/src/annotations.ts [hunk 3]
```diff
@@ -139,6 +139,14 @@ export const TOOL_ANNOTATIONS: Readonly<Record<string, ToolAnnotation>> = {
     windowedText: true,
   },
 
+  // --- feedback ---
+  // send_feedback carries a two-phase handshake for the email/stella channels;
+  // the CLI keeps it generic (no per-tool special-casing): the phase-1
+  // `approval_required` response renders as a single object showing
+  // `confirmation_token` and `next_step`, and the runtime prints a re-run hint
+  // driven off the response `status` field.
+  send_feedback: { command: ["feedback", "send"], scope: "feedback" },
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply packages/cli/src/annotations.ts:148]: Accepted. Added stella:feedback to the CLI known-scope allowlist and covered it with a small auth-constants test.

CC on behalf of @jan-kubica

File: apps/api/src/handlers/feedback/intake.ts [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    .digest("hex");
+
+const composeDeliveryBody = (
+  sanitizedBody: string,
+  source: FeedbackSource | undefined,
+): string => {
+  const sourceLine =
+    source === undefined
+      ? ""
+      : `\n_Source: instance=${source.instance ?? "unknown"}, version=${source.version ?? "unknown"}._`;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/feedback/intake.ts:208]: Accepted. Source metadata is now sanitized before it is appended to the delivery body or included in the private intake reporter block, with a regression test covering redacted instance/version fields.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/research-admin-tools.ts [hunk 5]
```diff
@@ -711,6 +722,20 @@ const handleManageOrganizationTool: McpToolHandler = async ({
...
+    // as a whole (it also adds members and updates settings), so the central
+    // gate in tools.ts cannot cover this. Removing a member is irreversible,
+    // so refuse until the human-approved `confirm: true` arrives (the CLI's
+    // --yes flow injects it). Mirrors the central gate's wording.
+    if (input.confirm !== true) {
+      return structuredErrorResult({
+        code: "confirmation_required",
+        message:
+          "remove_member is an irreversible operation and was called without confirmation",
+        hint: "Removing a member is irreversible. Confirm with the human user, then retry with confirm: true.",
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/research-admin-tools.ts:735]: Accepted. The chat write adapter now injects confirm: true for approved manage_organization/remove_member calls before dispatching to the MCP handler, with a focused unit assertion for that adapter behavior.\n\nCC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-token.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+};
+
+/** SHA-256 of the exact content the token authorizes, hex-encoded. */
+const contentHash = ({
+  kind,
+  sanitizedBody,
+  sanitizedTitle,
+}: FeedbackTokenContent): string =>
+  createHash("sha256")
+    .update(`${kind}\n${sanitizedTitle}\n${sanitizedBody}`)
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-token.ts]: Accepted. Feedback token content is now hashed from a structured JSON tuple rather than newline concatenation, and the tests cover the title/body boundary case directly.\n\nCC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.ts [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+  // Phase-2 delivery rate limit: bound actual sends per organization. Only the
+  // email and stella channels reach here (github returns above without any
+  // server-side delivery, and phase-1 previews already returned), so this
+  // counts deliveries, not previews.
+  const withinDeliveryLimit = await feedbackIntakeGuards.consumeCounter({
+    bucket: FEEDBACK_DELIVERY_BUCKET,
+    key: context.organizationId,
+    windowMs: FEEDBACK_DELIVERY_WINDOW_MS,
+    max: FEEDBACK_DELIVERY_MAX_PER_ORG,
+  });
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.ts:398]: Accepted. The phase-2 delivery counter remains a pre-send reservation, but failed delivery outcomes now release the slot, so feature-disabled config and retryable intake/email failures no longer spend the hourly delivery quota.\n\nCC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.ts [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+  const response = await Result.tryPromise({
+    try: async () =>
+      await fetch(intakeUrl, {
+        method: "POST",
+        headers: { "content-type": "application/json" },
+        body: JSON.stringify({
+          kind,
+          title: sanitizedTitle,
+          body: composedBody,
+          source: { instance: env.isDev ? "dev" : "self-hosted" },
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.ts]: Accepted. The stella channel now truncates the approved/forwarded body to the intake body cap before minting the confirmation token, so the human-approved preview and the forwarded payload are the same bounded content.\n\nCC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.ts [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+      code: "validation_error",
+      message: "This feedback was already forwarded to the stella intake",
+      hint: "An identical report was submitted recently; no need to resend.",
+    });
+  }
+  if (!response.ok) {
+    return structuredErrorResult({
+      code: "internal_error",
+      message: "The stella feedback intake rejected the submission",
+      hint: MCP_INTERNAL_ERROR_HINT,
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.ts:671]: Accepted and implemented. The stella forwarding path now preserves the intake feature_disabled envelope instead of converting it to a retryable internal error.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.ts [hunk 10]
```diff
@@ -0,0 +1,566 @@
...
+}: {
+  body: string;
+  kind: FeedbackKind;
+}): string =>
+  `${body}\n\n---\n_Filed via stella send_feedback (agent-assisted, sanitized). Kind: ${kind}._`;
+
+const fitStellaIntakeBody = (body: string): string => {
+  const bodyChars = Array.from(body);
+  if (bodyChars.length <= MAX_FEEDBACK_BODY_CHARS) {
+    return body;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.ts]: Accepted and implemented. Stella body fitting now uses the same JavaScript string-length metric as the intake schema and slices through the existing surrogate guard.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-token.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+const FEEDBACK_TOKEN_TTL_MS = FEEDBACK_TOKEN_TTL_MINUTES * 60 * 1000;
+
+export type FeedbackTokenContent = {
+  kind: string;
+  sanitizedTitle: string;
+  sanitizedBody: string;
+};
+
+const serializeContent = (content: FeedbackTokenContent): string =>
+  JSON.stringify([content.kind, content.sanitizedTitle, content.sanitizedBody]);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-token.ts]: Accepted and implemented. Feedback confirmation tokens now include the approved channel in the signed content, and send_feedback verifies the same channel on phase 2.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.ts [hunk 12]
```diff
@@ -0,0 +1,639 @@
...
+  }
+
+  const response = await Result.tryPromise({
+    try: async () =>
+      await fetch(intakeUrl, {
+        method: "POST",
+        headers: { "content-type": "application/json" },
+        body: JSON.stringify({
+          kind,
+          title: sanitizedTitle,
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.ts]: Accepted and implemented. Stella-channel feedback now caps the sanitized title to the intake title limit before token creation and phase-2 delivery, so the approved content is the exact content forwarded to the intake.

CC on behalf of @jan-kubica

File: apps/api/.env.example [hunk 13]
```diff
@@ -45,6 +45,16 @@ EMAIL_PROVIDER="smtp"
 # email. The domain must be verified with the chosen provider.
 TRANSACTIONAL_EMAIL_FROM="noreply@example.com"
 
+# Optional. Destination for the MCP `send_feedback` email channel and the public
+# feedback intake (POST /public/feedback). When unset, the tool only offers the
+# github channel (email/stella refuse with `feature_disabled`) and the intake
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/.env.example]: Fixed the example env guidance to separate local email/public-intake delivery from the `stella` channel forwarding target. It now notes that `stella` can still forward to `FEEDBACK_INTAKE_URL` when that is configured.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.ts [hunk 14]
```diff
@@ -0,0 +1,655 @@
...
+  kind,
+  sanitizedTitle,
+}: {
+  composedBody: string;
+  context: McpRequestContext;
+  kind: FeedbackKind;
+  sanitizedTitle: string;
+}): Promise<CallToolResult> => {
+  const feedbackEmailTo = env.FEEDBACK_EMAIL_TO;
+  if (!feedbackEmailTo) {
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.ts:442]: Fixed the email channel to return `feature_disabled` when transactional email transport is not configured, before calling `sendFeedbackEmail`. The new test covers `FEEDBACK_EMAIL_TO` being set while transport config is absent.

CC on behalf of @jan-kubica

File: apps/api/src/handlers/feedback/intake.ts [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+  body,
+  kind,
+  title,
+}: {
+  body: string;
+  kind: string;
+  title: string;
+}): string =>
+  new Bun.CryptoHasher("sha256")
+    .update(`${kind}\n${title}\n${body}`)
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/feedback/intake.ts]: Fixed the intake dedup key to hash JSON-serialized named fields instead of newline-joining the values. Added a regression test for the title/body newline boundary collision.

CC on behalf of @jan-kubica

File: apps/api/src/handlers/feedback/intake.ts [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+  title,
+}: {
+  composedBody: string;
+  emailTo: string | undefined;
+  kind: PublicFeedbackBody["kind"];
+  source: FeedbackSource | undefined;
+  title: string;
+}): Promise<DeliveryOutcome> => {
+  if (emailTo) {
+    return await deliverViaEmail({
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/feedback/intake.ts:283]: Accepted and implemented. The public intake now checks `isTransactionalEmailConfigured()` before calling `sendFeedbackEmail`, returning `feature_disabled` when `FEEDBACK_EMAIL_TO` exists but the transport settings are incomplete. Added a regression test that verifies the email sender is not called in that state.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.ts [hunk 17]
```diff
@@ -0,0 +1,665 @@
...
+    return structuredErrorResult({
+      code: "feature_disabled",
+      message: "Email feedback is not configured on this server",
+      hint: 'Email feedback is not configured on this server; use channel "github" or "stella".',
+    });
+  }
+  if (!isTransactionalEmailConfigured()) {
+    return structuredErrorResult({
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.ts:450]: Accepted and implemented at the shared guard. `isTransactionalEmailConfigured()` now treats partial SMTP and SES credential pairs as incomplete, so both the MCP email channel and public intake return `feature_disabled` before constructing a transport. Added direct tests for no-credential, complete-pair, and partial-pair cases.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-sanitize.ts [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+const REDACTED_URL = "[redacted-url]";
+const REDACTED_IP = "[redacted-ip]";
+
+/**
+ * URL hosts kept verbatim (all others collapse to `[redacted-url]`). Public,
+ * non-tenant destinations only: the code repo (where issues are filed) and
+ * Stella's public sites. A host matches when it equals a listed domain or is a
+ * subdomain of one.
+ */
+const URL_HOST_ALLOWLIST = ["github.com", "stella.legal", "stll.app"] as const;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-sanitize.ts]: Accepted and tightened. The sanitizer no longer preserves all GitHub-hosted URLs: only queryless, fragmentless `github.com/stella/stella` paths survive, private GitHub repo URLs are redacted, and broad `stll.app` preservation was narrowed to the public feedback intake path. Added regression coverage for private GitHub, GitHub query data, tenant app URLs, and the public intake URL.

CC on behalf of @jan-kubica

File: apps/api/src/handlers/feedback/routes.ts [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+ * strict "reject anything else" contract runs on the raw payload in the handler
+ * (Valibot `strictObject`), the same shape as the hosted-usage webhook. The
+ * route keeps only a coarse size cap as defense-in-depth before parsing.
+ */
+
+import { panic } from "better-result";
+import Elysia, { t } from "elysia";
+
+import {
+  MAX_RAW_FEEDBACK_BODY_BYTES,
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/feedback/routes.ts]: Accepted by making the contract match the implementation. The route-level cap is a text-string character cap, so `MAX_RAW_FEEDBACK_BODY_BYTES` is now `MAX_RAW_FEEDBACK_BODY_CHARS` and the surrounding comments describe the raw string cap instead of implying a UTF-8 byte bound.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.test.ts [hunk 20]
```diff
@@ -0,0 +1,760 @@
...
+      body: "The cursor docs are unclear.",
+      channel: "email",
+    };
+    const token = String(parsePayload(await send(args)).confirmation_token);
+
+    const result = await send({ ...args, confirmation_token: token });
+    const payload = parseErrorPayload(result);
+    expect(payload.error?.code).toBe("feature_disabled");
+    expect(sendFeedbackEmailMock).not.toHaveBeenCalled();
+  });
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.test.ts:770]: Accepted. The two email `feature_disabled` tests now assert the preview phase does not consume quota, and the refused delivery phase consumes once then releases once.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/README.md [hunk 21]
```diff
@@ -77,9 +77,129 @@ The OpenAI-compatible `search`/`fetch` tools keep their bespoke
...
+  and a token; a second call echoing the same content plus the token sends the
+  email via `sendFeedbackEmail` to `FEEDBACK_EMAIL_TO`. When that env var is
+  unset the channel refuses with `feature_disabled`. The email carries the
+  sanitized title/body plus a private reporter block (userId, organizationId,
+  and, best-effort, the user's email) that is never part of the public issue.
+- **stella (last resort):** for a human with no GitHub account on a server with
+  no maintainer inbox. Same confirmation-token handshake as email (the token
+  hash covers only kind/title/body, so it is channel-agnostic); on phase two the
+  sanitized payload is POSTed to the hosted intake at `FEEDBACK_INTAKE_URL`
+  (default `https://api.stll.app/public/feedback`, overridable for self-hosting).
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/README.md:152]: Accepted. The MCP feedback README now describes the token as channel-bound and notes that a token minted for one delivery channel cannot be replayed against another.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/server.test.ts [hunk 22]
```diff
@@ -230,14 +259,63 @@ describe("handleMcpHttpRequest", () => {
...
+              message:
+                "Insufficient permissions. Required scope: stella:skills",
+              hint: "Grant the 'stella:skills' scope by re-running OAuth consent (CLI: 'stella auth login'), then retry.",
+            },
+          }),
         },
       ],
       isError: true,
     });
   });
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/server.test.ts:275]: Accepted. The missing-scope hint now includes `stella auth login --scopes ${scope}` in the implementation, and the server test expectation covers the `stella:skills` case. The initialize instructions were updated to the same scoped recovery form.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-sanitize.ts [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+// Long base64url blob (>= 40 chars): opaque access tokens, secrets. The
+// base64url alphabet (no `+` or `/`) is used on purpose: including `/` would
+// let this pass swallow whole URL path segments, and modern tokens (GitHub
+// PATs, JWT parts, most API keys) are base64url anyway. A hex secret is caught
+// by HEX_SECRET_REGEX above.
+const BASE64_SECRET_REGEX = /\b[A-Za-z0-9_-]{40,}={0,2}/gu;
+
+// Absolute http(s) URL. Trailing sentence punctuation is trimmed in the
+// replacer so a URL at the end of a sentence keeps the period.
+const URL_REGEX = /\bhttps?:\/\/[^\s<>()[\]"'`]+/giu;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-sanitize.ts]: Accepted and implemented. The URL matcher now includes valid parenthesized/bracketed path text, then trims only trailing sentence punctuation and unmatched closing wrappers before parsing. Added coverage for an internal URL containing `(Acme-Secret)` and for a wrapped URL so the wrapper punctuation stays outside the redaction.

CC on behalf of @jan-kubica

File: apps/api/src/env.ts [hunk 24]
```diff
@@ -150,6 +150,25 @@ const envApi = createEnv({
...
+     * Where the MCP `send_feedback` "stella" channel forwards approved,
+     * sanitized feedback: the hosted intake receiver. Defaults to Stella's
+     * hosted API (`https://api.stll.app/public/feedback`); a self-hosted
+     * deployment can point it at its own intake or a proxy. When the tool runs
+     * on the hosted instance this points back at itself, which is a normal
+     * round trip.
+     */
+    FEEDBACK_INTAKE_URL: v.optional(
+      v.pipe(v.string(), v.url()),
+      "https://api.stll.app/public/feedback",
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/env.ts:170]: Accepted and implemented. Hosted `channel: "stella"` delivery now routes to the intake handler in-process when `FEEDBACK_INTAKE_URL` is the hosted default, with the public IP limiter skipped for that trusted self-forward. The MCP per-org delivery limiter and intake dedup/delivery validation still run. Custom intake URLs still use `fetch`. Added coverage for the hosted internal path.

CC on behalf of @jan-kubica

File: apps/api/src/lib/email/index.tsx [hunk 25]
```diff
@@ -198,3 +195,40 @@ export const sendOrganizationInvitation = async ({
...
   });
 };
+
+type SendFeedbackEmailProps = {
+  to: string;
+  kind: ProductFeedbackKind;
+  title: string;
+  body: string;
+  reporter: ProductFeedback.FeedbackReporter;
+};
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/email/index.tsx]: Accepted. Renamed the non-React function parameter type from `SendFeedbackEmailProps` to `SendFeedbackEmailOptions`; the shape and call sites are unchanged.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/feedback-tools.ts [hunk 26]
```diff
@@ -0,0 +1,687 @@
...
+    source: { instance: env.isDev ? "dev" : "self-hosted" },
+  });
+
+  if (isHostedFeedbackIntakeUrl(intakeUrl)) {
+    return await mapIntakeResponse(
+      await receivePublicFeedback({
+        rawBody,
+        clientIp: null,
+        deps: { skipRateLimit: true },
+      }),
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/feedback-tools.ts]: Accepted. The internal delivery shortcut now requires both the hosted intake URL and this server public origin to be the hosted API origin. A self-hosted instance left on the default `https://api.stll.app/public/feedback` still posts over HTTP to the hosted intake. Added coverage for that default self-hosted path.

CC on behalf of @jan-kubica

File: apps/api/src/handlers/feedback/intake.ts [hunk 27]
```diff
@@ -0,0 +1,375 @@
...
+const FEEDBACK_KINDS = ["bug", "feature_request", "docs", "other"] as const;
+
+const MAX_TITLE_CHARS = 200;
+const MAX_BODY_CHARS = 8000;
+const MAX_SOURCE_FIELD_CHARS = 40;
+
+// Coarse raw-body string cap enforced at the route before JSON parsing: bounds
+// an unauthenticated write independently of the per-field caps below. Sized for
+// an 8000-char body plus title, source, and JSON overhead.
+export const MAX_RAW_FEEDBACK_BODY_CHARS = 16_384;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/feedback/intake.ts]: Accepted. Increased the unauthenticated raw intake cap to cover max-length valid fields after JSON escaping, while keeping the parsed Valibot limits unchanged. Added route-level coverage for a valid max-length payload whose serialized JSON expands through escaping.

CC on behalf of @jan-kubica

File: apps/api/src/mcp/server-core.ts [hunk 28]
```diff
@@ -167,37 +180,29 @@ export const createMcpHttpRequestHandler = ({
...
-            { type: "text" as const, text: `Unknown tool: ${toolName}` },
-          ],
-          isError: true,
-        };
+        // Suggest the closest names the caller can actually see (scope-filtered
+        // list), so a typo resolves without leaking tools they lack access to.
+        const visibleNames = (
+          await listMcpTools(context, mode, session.scopes)
+        ).map((tool) => tool.name);
+        const suggestions = closestToolNames(toolName, visibleNames);
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/mcp/server-core.ts]: Accepted. Unknown tool suggestions now run only for names within the normal tool-name bound, and long unknown names are truncated in the error message. Long malformed names still return the structured `unknown_tool` envelope, but skip the scope-filtered visible-tool listing and fuzzy distance work. Added coverage for that path.

CC on behalf of @jan-kubica
